src/hotspot/share/runtime/stackWatermarkSet.cpp

```diff
@@ -0,0 +1,145 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "logging/log.hpp"
+#include "runtime/atomic.hpp"
+#include "runtime/frame.inline.hpp"
+#include "runtime/safepoint.hpp"
+#include "runtime/safepointMechanism.inline.hpp"
+#include "runtime/stackWatermark.inline.hpp"
+#include "runtime/stackWatermarkSet.inline.hpp"
+#include "runtime/thread.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/preserveException.hpp"
+#include "utilities/vmError.hpp"
+
+StackWatermarks::StackWatermarks() :
+    _head(NULL) {}
+
+StackWatermarks::~StackWatermarks() {
+  StackWatermark* current = _head;
+  while (current != NULL) {
+    StackWatermark* next = current->next();
+    delete current;
+    current = next;
+  }
+}
+
+StackWatermark* StackWatermarkSet::head(JavaThread* jt) {
+  return jt->stack_watermarks()->_head;
+}
+
+void StackWatermarkSet::set_head(JavaThread* jt, StackWatermark* watermark) {
+  jt->stack_watermarks()->_head = watermark;
+}
+
+void StackWatermarkSet::add_watermark(JavaThread* jt, StackWatermark* watermark) {
+  assert(!has_watermark(jt, watermark->kind()), "Two instances of same kind");
+  StackWatermark* prev = head(jt);
+  watermark->set_next(prev);
+  set_head(jt, watermark);
+}
+
+static void verify_poll_context() {
+#ifdef ASSERT
+  Thread* thread = Thread::current();
+  if (thread->is_Java_thread()) {
+    JavaThread* jt = static_cast<JavaThread*>(thread);
+    JavaThreadState state = jt->thread_state();
+    assert(state != _thread_in_native, "unsafe thread state");
+    assert(state != _thread_blocked, "unsafe thread state");
+  } else if (thread->is_VM_thread()) {
+  } else {
+    assert_locked_or_safepoint(Threads_lock);
+  }
+#endif
+}
+
+void StackWatermarkSet::before_unwind(JavaThread* jt) {
+  verify_poll_context();
+  if (!jt->has_last_Java_frame()) {
+    // Sometimes we throw exceptions and use native transitions on threads that
+    // do not have any Java threads. Skip those callsites.
+    return;
+  }
+  for (StackWatermark* current = head(jt); current != NULL; current = current->next()) {
+    current->before_unwind();
+  }
+  SafepointMechanism::update_poll_values(jt);
+}
+
+void StackWatermarkSet::after_unwind(JavaThread* jt) {
+  verify_poll_context();
+  if (!jt->has_last_Java_frame()) {
+    // Sometimes we throw exceptions and use native transitions on threads that
+    // do not have any Java threads. Skip those callsites.
+    return;
+  }
+  for (StackWatermark* current = head(jt); current != NULL; current = current->next()) {
+    current->after_unwind();
+  }
+  SafepointMechanism::update_poll_values(jt);
+}
+
+void StackWatermarkSet::on_iteration(JavaThread* jt, const frame& fr) {
+  if (VMError::is_error_reported()) {
+    // Don't perform barrier when error reporting walks the stack.
+    return;
+  }
+  verify_poll_context();
```

Comment:
There's a verfy_poll_context here, but no update_poll_values call. I guess this is because we could be iterating over a thread that is not Thread::current()? But in that case, should we really be "verifying the poll context" of the current thread?

Reply:
Correct - I'm not updating the poll values, because it's not necessarily the current thread we are processing. But the verify_poll_context() code does correctly always refer to the current frame. It verifies we are in the right context to perform this kind of processing, yet the processing could be for a different thread.

Reply:
Updating the name to verify_processing_context to better reflect what this does.